### ISSUE TITLE
Distributed UNION ALL exception

### ISSUE THREAD

**Issue by jaykelin** [yandex/ClickHouse#1059]
execute SQL like that
`select b, sum(c) from ( select a, b, sum(c) as c from distributed.table_a where ... group by a,b UNION ALL select a, b, sum(c) as c from distributed.table_b where ... group by a, b ) as a group by b`
Same SQL but change table to **MergeTree** is ok.

Distributed UNION ALL catch exception like 

> ru.yandex.clickhouse.except.ClickHouseException: ClickHouse exception, code: 11, host: host_name, port: 8123; Code: 11, e.displayText() = DB::Exception: Position 2 is out of bound in Block::safeGetByPosition(), max position = 1, there are columns: etldate, if(equals(tv, 'all'), '汇总', tv), e.what() = DB::Exception
> 
>         at ru.yandex.clickhouse.except.ClickHouseExceptionSpecifier.specify(ClickHouseExceptionSpecifier.java:57)
>         at ru.yandex.clickhouse.except.ClickHouseExceptionSpecifier.specify(ClickHouseExceptionSpecifier.java:28)
>         at ru.yandex.clickhouse.ClickHouseStatementImpl.getInputStream(ClickHouseStatementImpl.java:445)
>         at ru.yandex.clickhouse.ClickHouseStatementImpl.executeQuery(ClickHouseStatementImpl.java:85)
>         at ru.yandex.clickhouse.ClickHouseStatementImpl.executeQuery(ClickHouseStatementImpl.java:81)

**Comment by proller** [comment]
@jaykelin Can you prove some repeatable test?
Usually it contain queries: create table; inssert; select;
Now we have 
https://github.com/yandex/ClickHouse/blob/master/dbms/tests/queries/0_stateless/00506_union_distributed.sql but it always pass, before and after your changes

**Comment by lppsuixn** [comment]
@proller 
Example:
```sql
CREATE TABLE test_local.union_all_bug on cluster bi_cluster (etldate Date  DEFAULT CAST(0 AS Int64),v1 Int64,v2 Int64  DEFAULT CAST(0 AS Int64)) ENGINE =ReplicatedMergeTree('/clickhouse/tables/test/union_all_bug/{shard}', '{replica}', etldate, (etldate), 8192);

CREATE TABLE test.union_all_bug  (etldate Date  DEFAULT CAST(0 AS Int64),v1 Int64,v2 Int64  DEFAULT CAST(0 AS Int64)) ENGINE =  Distributed(bi_cluster, 'test_local', 'union_all_bug',cityHash64(etldate)) ;
```

Then:

```sql
insert into test.union_all_bug values('2017-11-02',3,4);
insert into test.union_all_bug values('2017-11-01',1,2);
```
bellow query not work:

```sql
select etldate,sum(v1) from (select * from ( select etldate,v1,v2 as v3 from test.union_all_bug union all select etldate,sum(v1) as v1,sum(v2) as v3 from test.union_all_bug group by etldate )) AS __view__ GROUP BY etldate;

Received exception from server:
Code: 11. DB::Exception: Received from clickhouse.bi.kgidc.cn:9000, 10.1.169.198. DB::Exception: Position 2 is out of bound in Block::safeGetByPosition(), max position = 1, there are columns: etldate, sum(v1). Stack trace:
```
and this is working fine:
```sql
select etldate,sum(v1),sum(v3) from (select * from ( select etldate,v1,v2 as v3 from test.union_all_bug union all select etldate,sum(v1) as v1,sum(v2) as v3 from test.union_all_bug group by etldate )) AS __view__ GROUP BY etldate;

┌────etldate─┬─sum(v1)─┬─sum(v3)─┐
│ 2017-11-01 │       2 │       4 │
│ 2017-11-02 │       6 │       8 │
└────────────┴─────────┴─────────┘
```

**Comment by jaykelin** [comment]
@proller How is the case test going? 
we has merge request #1233 to 1.1.54310-stable in our production cluster, i hope this request will be merged in the next release.

**Comment by zhang2014** [comment]
This problem is the same as #320, which is caused by `rewriteExpressionList` in subQuery.I've fixed it in #1763. Can we close this issue? @proller
